README.md

## Introduction

My first introduction to running a command as a daemon looked like this:

    nohup mycommand --myoption myarg >/dev/null 2>/dev/null &

While this is a good for doing some development work, a proper daemon requires
quite a number of other things to happen.  The wikipedia page
http://en.wikipedia.org/wiki/Daemon\_(computing) says the following needs to
happen:

- Dissociating from the controlling tty
- Becoming a session leader
- Becoming a process group leader
- Executing as a background task by forking and exiting (once or twice). This
  is required sometimes for the process to become a session leader. It also
  allows the parent process to continue its normal execution.
- Setting the root directory ("/") as the current working directory so that the
  process does not keep any directory in use that may be on a mounted file
  system (allowing it to be unmounted).
- Changing the umask to 0 to allow open(), creat(), et al. operating system
  calls to provide their own permission masks and not to depend on the umask of
  the caller
- Closing all inherited files at the time of execution that are left open by
  the parent process, including file descriptors 0, 1 and 2 (stdin, stdout,
  stderr). Required files will be opened later.
- Using a logfile, the console, or /dev/null as stdin, stdout, and stderr

Additionally, when running a process under monit or initd, it is typical for
the process id of the daemon to be stored in a file, so the daemon can be
monitored and killed easily.

Finally, it is common to use `start`, `stop`, `restart`, and
`status` commands to control the daemon.

Daemoncmd takes care of creating a daemon the right way, putting its pid in
a file, and interacting with the daemon via `start`, `stop`, etc. commands.
Using it is as simple as:

    /path/to/daemoncmd start --pidfile /tmp/mydaemon.pid /path/to/mycommand

See the Usage section for more details.

For more information on daemons and daemons in python, see:

- http://pypi.python.org/pypi/python-daemon
- http://www.python.org/dev/peps/pep-3143/
- http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/66012
- http://en.wikipedia.org/wiki/Daemon\_(computing)


## Alternatives

Here are some projects that are similar in scope to daemoncmd:

- http://pypi.python.org/pypi/zdaemon/2.0.4, and see
  http://ridingpython.blogspot.cz/2011/08/turning-your-python-script-into-linux.html
  for a description.
- https://github.com/indexzero/forever

While daemoncmd is a great way to start running a daemon, there are many
programs designed to run and monitor processes in a way that is more robust and
full-featured than simply starting your own daemon.  Here is a list of some of
them (see http://news.ycombinator.com/item?id=1368855 for a discussion):

- launchd http://launchd.macosforge.org/
- upstart http://upstart.ubuntu.com/
- monit http://mmonit.com/monit/
- supervisord http://supervisord.org/
- daemontools http://cr.yp.to/daemontools.html
- daemonize http://bmc.github.com/daemonize/
- runit http://smarden.sunsite.dk/runit/
- perp http://b0llix.net/perp/

For production systems, I recommend using a tool like one of these.  I have
used supervisord and monit before.  Some of these tools work with daemons, like
monit, in which case daemoncmd could come in handy.  Other tools, like
supervisord, can supervise your process in such a way that you do not need to
daemonize it.


## Contribute

If this project does not do what you want, please open an issue or a pull
request on github, https://github.com/todddeluca/daemoncmd.


## Requirements

- Probably Python 2.7 (since that is the only version it has been tested with.)


## Installation

### Install from pypi.python.org

Download and install python-vagrant:

    pip install daemoncmd

### Install from github.com

Clone and install daemoncmd:

    git clone git@github.com:todddeluca/daemoncmd.git
    cd daemoncmd
    python setup.py install


## Usage

This module has two separate uses cases: 
    
- running a command as a daemon process
- forking the current python process as a daemon.

Daemonizing a command allows one to start, stop, and restart a non-daemon
command as a daemon process.  This requires specifying a pid file which is used
to interact with the process.

Usage examples:

     daemoncmd start --pidfile /tmp/daemon.pid \
            --stdout /tmp/daemon.log --stderr /tmp/daemon.log sleep 100
     daemoncmd restart --pidfile /tmp/daemon.pid \
            --stdout /tmp/daemon.log --stderr /tmp/daemon.log sleep 100
     daemoncmd status --pidfile /tmp/daemon.pid
     daemoncmd stop --pidfile /tmp/daemon.pid

Another use case is forking the current process into a daemon.  According
to pep 3143, forking as a daemon might be done by the standard library some
day.

Usage example:

    import daemoncmd
    import mytask

    daemoncmd.daemonize()
    mytask.doit()

Or from the command line:

    python -c 'import daemoncmd, mytask; daemoncmd.daemonize(); mytask.doit()'

Other usage notes:

- The command should not daemonize itself, since that is what this script does
  and it would make the pid in the pidfile incorrect.
- The command should refer to the absolute path of the executable, since
  daemonization sets the current working directory (cwd) to '/'.  More
  generally, do not assume what the cwd is.
- If daemoncmd is run by monit, etc., PATH and other env vars might be
  restricted for security reasons.
- daemoncmd does not try to run the daemon as a particular uid.  That would
  be handled by a process manager like monit, launchd, init, god, etc.
- When running under monit, etc., pass environment variables to the command
  like so:

        FOO=testing daemoncmd start --pidfile /tmp/daemon.pid \
            --stdout /tmp/daemon.log printenv FOO







CHANGELOG.md
# Changelog

This document lists the changes (and individuals who contributed to those
changes) for each release.


## 0.2.0 (released 2013/06/26)

- Include `__version__` in daemoncmd.py module.
- Replace `bin/daemoncmd` with a setuptools `console_scripts` entry point.
- Update tests to reflect that the CLI main function has moved from
  `bin/daemoncmd` to `daemoncmd.py`.
- Add this change log.



tests/test_daemoncmd.py




import subprocess
import os
import sys


PID_FILE = os.path.abspath('test_daemon.pid')
LOG_FILE = os.path.abspath('test_daemon.log')

# assume this directory structure
# project/tests/test_daemoncmd.py
# project/daemoncmd.py

HERE = os.path.dirname(os.path.abspath(__file__))
DC_PATH = os.path.join(os.path.dirname(HERE), 'daemoncmd.py')
DAEMONCMD = sys.executable + ' ' + DC_PATH

def test_cli():
    '''
    Test start, status, restart, and stop using the bin/daemoncmd command
    line interface.  The "test" is that the subprocess calls do not return a
    non-zero return code.
    '''
    kws = {'pid': PID_FILE, 'out': LOG_FILE, 'err': LOG_FILE}

    status_cmd = DAEMONCMD + ' status --pidfile {pid}'
    status_cmd = status_cmd.format(**kws)
    stop_cmd = DAEMONCMD + ' stop --pidfile {pid}'
    stop_cmd = stop_cmd.format(**kws)
    start_cmd = DAEMONCMD + ' start --pidfile {pid} --stdout {out}'
    start_cmd += ' --stderr {err} sleep 10'
    start_cmd = start_cmd.format(**kws)
    restart_cmd = DAEMONCMD + ' restart --pidfile {pid} --stdout {out}'
    restart_cmd += ' --stderr {err} sleep 10'
    restart_cmd = restart_cmd.format(**kws)
    try:
        # check the process is not running yet.
        output = subprocess.check_output(status_cmd, shell=True)
        print output.find('stopped')
        assert output.find('stopped') != -1

        # start process
        cmd = DAEMONCMD + ' start --pidfile {pid} --stdout {out}'
        cmd += ' --stderr {err} sleep 10'
        output = subprocess.check_output(start_cmd, shell=True)

        # check that process is running
        cmd = DAEMONCMD + ' status --pidfile {pid}'
        output = subprocess.check_output(status_cmd, shell=True)
        print output.find('running')
        assert output.find('running') != -1

        # restart process
        cmd = DAEMONCMD + ' restart --pidfile {pid} --stdout {out}'
        cmd += ' --stderr {err} sleep 10'
        output = subprocess.check_output(restart_cmd, shell=True)

        # check that process is running
        cmd = DAEMONCMD + ' status --pidfile {pid}'
        output = subprocess.check_output(status_cmd, shell=True)
        print output.find('running')
        assert output.find('running') != -1

        # stop process
        cmd = 'bin/daemoncmd stop --pidfile {pid}'
        output = subprocess.check_output(stop_cmd, shell=True)

        # check that process is stopped
        cmd = 'bin/daemoncmd status --pidfile {pid}'
        output = subprocess.check_output(status_cmd, shell=True)
        print output.find('stopped')
        assert output.find('stopped') != -1
    finally:
        for path in (PID_FILE, LOG_FILE):
            if os.path.exists(path):
                os.remove(path)





daemoncmd.py
#!/usr/bin/env python2.7

'''
For more information on daemons in python, see:

* http://pypi.python.org/pypi/python-daemon
* http://www.python.org/dev/peps/pep-3143/
* http://aspn.activestate.com/ASPN/Cookbook/Python/Recipe/66012
* http://en.wikipedia.org/wiki/Daemon_(computing)

Here are some similar implementations:

* http://pypi.python.org/pypi/zdaemon/2.0.4
* https://github.com/indexzero/forever

This module has two separate uses cases: 
    
* running a command as a daemon process
* forking the current process as a daemon.

Daemonizing a command allows one to start, stop, and restart a non-daemon
command as a daemon process.  This requires specifying a pid file which is used
to interact with the process.

Usage examples:

    daemoncmd start --pidfile /tmp/daemon.pid \
            --stdout /tmp/daemon.log --stderr /tmp/daemon.log sleep 100
    daemoncmd restart --pidfile /tmp/daemon.pid \
            --stdout /tmp/daemon.log --stderr /tmp/daemon.log sleep 100
    daemoncmd status --pidfile /tmp/daemon.pid
    daemoncmd stop --pidfile /tmp/daemon.pid

Another use case is forking the current process into a daemon.  According
to pep 3143, forking as a daemon might be done by the standard library some
day.

Usage example:

    import daemoncmd
    import mytask

    daemoncmd.daemonize()
    mytask.doit()

Or from the command line:

    python -c 'import daemoncmd, mytask; daemoncmd.daemonize(); mytask.doit()'

Other usage notes:

* The command should not daemonize itself, since that is what this script does
  and it would make the pid in the pidfile incorrect.
* The command should be refer to the absolute path of the executable, since
  daemonization sets the cwd to '/'.  More generally, do not assume what the
  cwd is.
* If daemoncmd is run by monit, etc., PATH and other env vars might be
  restricted for security reasons.
* daemoncmd does not try to run the daemon as a particular uid.  That would
  be handled by a process manager like monit, launchd, init, god, etc.
* When running under monit, etc., pass environment variables to the command
  like so:

    FOO=testing daemoncmd start --pidfile /tmp/daemon.pid \
            --stdout /tmp/daemon.log printenv FOO
'''

__version__ = '0.2.0'

import sys
import os
import signal
import errno
import time
import argparse




def start(argv, pidfile, stdin='/dev/null', stdout='/dev/null',
          stderr='/dev/null'):
    '''
    Start a new daemon, saving its pid to pidfile.
    Do not start the daemon if the pidfile exists and the pid in it is running.
    '''
    # use absolute path, since daemonize() changes cwd
    pidfile = os.path.abspath(pidfile)
    pid = getpid(pidfile)

    # start process pidfile does not have pid or the pid is not a running
    # process.
    if pid and running(pid):
        mess = "Start aborted since pid file '%s' exists" % pidfile
        mess += " and pid '%s' is running.\n" % pid
        sys.stderr.write(mess)
        sys.exit(1)

    sys.stdout.write('Starting process.\n')
    daemonize_command(argv, pidfile, stdin, stdout, stderr)


def stop(pidfile):
    '''
    pidfile: a file containing a process id.
    stop the pid in pidfile if pidfile contains a pid and it is running.
    '''
    # use absolute path, since daemonize() changes cwd
    pidfile = os.path.abspath(pidfile)
    pid = getpid(pidfile)

    # stop process (if it exists)
    if not pid:
        sys.stderr.write(("Warning: Could not stop process because pid file "
                          "'%s' is missing.\n" % pidfile))
    elif not running(pid):
        sys.stderr.write(('Warning: pid "%s" in pid file "%s" is already not '
                          'running.\n' % (pid, pidfile)))
    else:
        sys.stdout.write('Stopping process. pid={0}\n'.format(pid))
        try:
            os.kill(pid, signal.SIGTERM)
            # a pause, so daemon will have a chance to stop before it gets restarted.
            time.sleep(1) 
        except OSError as err:
            sys.stderr.write('Failed to terminate pid "%s".  Exception: %s.\n'
                             % (pid, err))
            sys.exit(1)


def restart(argv, pidfile, stdin='/dev/null', stdout='/dev/null',
            stderr='/dev/null'):
    '''
    stop the process in pidfile.  start argv as a new daemon process.  save its
    pid to pidfile.
    '''
    stop(pidfile)
    start(argv, pidfile, stdin, stdout, stderr)


def status(pidfile):
    # use absolute path, since daemonize() changes cwd
    pidfile = os.path.abspath(pidfile)
    pid = getpid(pidfile)
    if pid and running(pid):
        sys.stdout.write('process running; pid={0}\n'.format(pid))
    else:
        sys.stdout.write('process stopped\n')
        

def daemonize_command(argv, pidfile, stdin='/dev/null', stdout='/dev/null',
                      stderr='/dev/null'):
    '''
    argv: list of executable and arguments.  First item is the executable. e.g.
    ['/bin/sleep', '100']
    pidfile: filename.  pid of the daemon child process will be written to the
    file.  This is useful for monitoring services that need a pid to stop the
    daemon, etc.

    Calls daemonize, which exits the calling process and continues in the child
    process.  Therefore all code after calling daemonize_command will be
    executed in the daemon process.
    '''
    # use absolute path, since daemonize() changes cwd
    pidfile = os.path.abspath(pidfile)
    daemonize(stdin, stdout, stderr)
    # now we are in the daemon process

    # save pid to a file
    if pidfile:
        setpid(os.getpid(), pidfile)

    # do not spawn a subprocess, since the daemon process is the one we want to
    # start/stop/restart/etc.
    os.execvp(argv[0], argv)

    
def daemonize(stdin='/dev/null', stdout='/dev/null', stderr='/dev/null'):
    '''
    stdin, stdout, stderr: filename that will be opened and used to replace the
    standard file descriptors is sys.stdin, sys.stdout, sys.stderr.  Default to
    /dev/null.  Note that stderr is opened unbuffered, so if it shares a file
    with stdout then interleaved output may not appear in the order that you
    expect.

    Turn current process into a daemon.
    returns: nothing in particular.
    '''

    # use absolute path, since daemonize() changes cwd
    stdin = os.path.abspath(stdin)
    stdout = os.path.abspath(stdout)
    stderr = os.path.abspath(stderr)
    
    # Do first fork
    try: 
        pid = os.fork() 
    except OSError as e:
        sys.stderr.write("fork #1 failed: (%d) %s\n"%(e.errno, e.strerror))
        sys.exit(1)
    if pid > 0:
        sys.exit(0) # exit parent
        
    # Decouple from parent environment.
    os.chdir("/") 
    os.umask(0) 
    os.setsid() 

    # Do second fork.
    try: 
        pid = os.fork() 
    except OSError as e: 
        sys.stderr.write("fork #2 failed: (%d) %s\n"%(e.errno, e.strerror))
        sys.exit(1)
    if pid > 0:
        sys.exit(0) # exit parent

    # Now I am a daemon!
    
    # Redirect standard file descriptors. First open the new files, perform
    # hack if necessary, flush any existing output, and dup new files to std
    # streams.
    si = open(stdin, 'r')
    so = open(stdout, 'a+')
    se = open(stderr, 'a+', 0)

    # hack and bug: when sys.stdin.close() has already been called,
    # os.dup2 throws an exception: ValueError: I/O operation on closed file
    # This hack attempts to detect whether any of the std streams
    # have been closed and if so opens them to a dummy value which
    # will get closed by os.dup2, which I like better than
    # an exception being thrown.
    if sys.stdin.closed: sys.stdin = open('/dev/null', 'r')
    if sys.stdout.closed: sys.stdout = open('/dev/null', 'a+')
    if sys.stderr.closed: sys.stderr = open('/dev/null', 'a+')
    
    sys.stdout.flush()
    sys.stderr.flush()

    os.dup2(si.fileno(), sys.stdin.fileno())
    os.dup2(so.fileno(), sys.stdout.fileno())
    os.dup2(se.fileno(), sys.stderr.fileno())


def getpid(pidfile):
    '''
    read pid from pidfile
    return: pid
    '''
    pid = None
    if os.path.isfile(pidfile):
        with open(pidfile) as fh:
            try:
                pid = int(fh.read().strip())
            except ValueError:
                pass
    return pid


def setpid(pid, pidfile):
    '''
    save pid to pidfile
    '''
    with open(pidfile, 'w') as fh:
        fh.write('{0}\n'.format(pid))


def running(pid):
    """
    pid: a process id
    Return: False if the pid is None or if the pid does not match a
    currently-running process.
    Derived from code in http://pypi.python.org/pypi/python-daemon/ runner.py
    """
    if pid is None:
        return False
    try:
        os.kill(pid, signal.SIG_DFL)
    except OSError, exc:
        if exc.errno == errno.ESRCH:
            # The specified PID does not exist
            return False
    return True


def main():    
    # daemoncmd <command>
    # daemoncmd start --pidfile <file> [--stdin <file>] [--stdout <file>] \
    #   [--stderr <file>] <command>
    # daemoncmd restart --pidfile <file> [--stdin <file>] [--stdout <file>]\
    #   [--stderr <file>] <command>
    # daemoncmd stop --pidfile <file>
    # daemoncmd status --pidfile <file>

    parser = argparse.ArgumentParser(
        description=('Turn any command into a daemon.  Use start, stop, ' +
                     'restart, and status to control the daemon.'))
    subparsers = parser.add_subparsers(dest='action')

    # create the parser for the "start" command
    startParser = subparsers.add_parser('start', 
                                        help='Start a daemon to run a command')
    startParser.add_argument(
        '--pidfile', required=True, 
        help='file in which to store the pid of the started daemon process')
    startParser.add_argument('--stdin', default='/dev/null', 
                             help='Redirect daemon stdin from this file')
    startParser.add_argument('--stdout', default='/dev/null', 
                             help='Redirect daemon stdout to this file')
    startParser.add_argument('--stderr', default='/dev/null', 
                             help='Redirect daemon stderr to this file')
    startParser.add_argument(
        'cmd', 
        help=('The executable/command that the daemon will run.  i.e. a '
              'server that listens on a port for incoming connections.'))
    startParser.add_argument('args', nargs='*', 
                             help='options or arguments to the command')
    
    stopParser = subparsers.add_parser('stop', help='Stop a running daemon')
    stopParser.add_argument('--pidfile', required=True, 
                            help='file containing the pid of daemon process')

    stopParser = subparsers.add_parser(
        'status', help='Print the status of a daemon process')
    stopParser.add_argument('--pidfile', required=True, 
                            help='file containing the pid of daemon process')

    restartParser = subparsers.add_parser(
        'restart', help='Restart a daemon to run a command')
    restartParser.add_argument(
        '--pidfile', required=True, 
        help='file in which to store the pid of the started daemon process')
    restartParser.add_argument('--stdin', default='/dev/null', 
                               help='Redirect daemon stdin from this file')
    restartParser.add_argument('--stdout', default='/dev/null', 
                               help='Redirect daemon stdout to this file')
    restartParser.add_argument('--stderr', default='/dev/null', 
                               help='Redirect daemon stderr to this file')
    restartParser.add_argument(
        'cmd', 
        help=('The executable/command that the daemon will run.  i.e. a '
              'server that listens on a port for incoming connections.'))
    restartParser.add_argument('args', nargs='*', 
                               help='options or arguments to the command')

    args = parser.parse_args()

    if args.action == 'start':
        start([args.cmd] + args.args, args.pidfile, args.stdin, args.stdout, args.stderr)
    elif args.action == 'restart':
        restart([args.cmd] + args.args, args.pidfile, args.stdin, args.stdout, args.stderr)
    elif args.action == 'stop':
        stop(args.pidfile)
    elif args.action == 'status':
        status(args.pidfile)


if __name__ == '__main__':
    main()

